components/remote_settings/src/storage.rs

```diff
@@ -88,14 +91,48 @@ impl Storage {
             .query_row(params![collection_url], |row| row.get(0))
             .optional()?;
 
+        let env_context = match env_context {
+            Some(ref ctx_str) => serde_json::from_str(ctx_str)?,
+            None => {
+                // Handle the None case
+                serde_json::Value::Null // or serde_json::json!({})
+            }
+        };
+
         let result = match fetched {
             Some(true) => {
-                // If fetched before, get the records from the records table
-                let records: Vec<RemoteSettingsRecord> = tx
-                    .prepare("SELECT data FROM records WHERE collection_url = ?")?
-                    .query_map(params![collection_url], |row| row.get::<_, Vec<u8>>(0))?
-                    .map(|data| serde_json::from_slice(&data.unwrap()).unwrap())
-                    .collect();
+                // Get all records for the collection_url
+                let mut stmt = tx.prepare(
+                    "SELECT data, filter_expression FROM records WHERE collection_url = ?",
+                )?;
+                let mut rows = stmt.query(params![collection_url])?;
+
+                let mut records = Vec::new();
+                while let Some(row) = rows.next()? {
+                    let data: Vec<u8> = row.get(0)?;
+                    let filter_expr: String = row.get(1)?;
+                    let record: RemoteSettingsRecord = serde_json::from_slice(&data)?;
+
+                    // Evaluate the filter_expression
+                    match evaluate_filter_expression(&filter_expr, &env_context) {
+                        Ok(true) => {
+                            // Include the record
+                            records.push(record);
+                        }
+                        Ok(false) => {
+                            // Exclude the record
+                            continue;
+                        }
+                        Err(e) => {
```

Comment:
Not sure what should happen in an error case here. Abort or just logging?

Reply:
Because they are target filters, not exclude filters, if the expressions fails to be evaluated then we consider the records has to be excluded.